v4/newrelic/application.go

```diff
@@ -120,5 +151,49 @@ func NewApplication(opts ...ConfigOption) (*Application, error) {
 			propagation.WithInjectors(trace.TraceContext{}),
 			propagation.WithExtractors(trace.TraceContext{}))
 	}
-	return &Application{tracer: tracer, propagators: propagators}, nil
+
+	if c.Logger == nil {
+		c.Logger = logger.ShimLogger{}
+	}
+	msg := fmt.Sprintf("Starting New Relic Shim Agent. If everything is configured properly, you should be able"+
+		" to see your data in a few minutes by visiting https://one.newrelic.com/launcher/nr1-core.explorer and"+
+		" searching for '%s'.", c.AppName)
+	c.Logger.Info(msg, nil)
+
+	return &Application{
+		tracer:      tracer,
+		propagators: propagators,
+		logger:      c.Logger,
+	}, nil
+}
+
+// LogInfo logs the given information at INFO level
+func (app *Application) LogInfo(msg string, context map[string]interface{}) {
+	if app == nil {
+		return
+	}
+	app.initLogger()
+	app.logger.Info(msg, context)
+}
+
+// LogDebug logs the given information at DEBUG level
+func (app *Application) LogDebug(msg string, context map[string]interface{}) {
+	if app == nil {
+		return
+	}
+	app.initLogger()
+	app.logger.Debug(msg, context)
+}
+
+func (app *Application) initLogger() {
+	if app.logger == nil {
+		app.logger = logger.ShimLogger{}
+	}
+}
+
+func logUnimplemented(logger Logger, methodName string) {
```

Comment:
Why is this not a method on `app` like the other logging methods?

Reply:
The answer to this one is also the answer to the previous question. Long story short, I was trying to repeat code as little as possible, and I used this `logUnimplemented` method in a few different places, including on the `Application`, but also on the `Transaction`. I didn't want to check the `app.logger` more times than I need to. But since you brought up the question I am realizing that this isn't very readable and is probably more prone to bugs. I'll try to improve it.

Reply:
Not sure if this is where you are headed with your refactor... Mentioning it to provide extra input in case it helps. When reading through the logging changes, I am left wondering why the majority of the logging functions are being added to `Application` instead of `Logger`? All the call sites seem to have access to a reference to a `Logger`. Not all of them have access to an `Application`.

Reply:
Originally I added them to the `Logger`, but the issue I ran into was that someone could conceivably manually create an `Application` object (rather than calling `NewApplication`), in which case the `Logger` on the app would be `nil`, which could lead to panics if we don't do a nil check each time. The idea was to add methods on the app so that we could keep the nil checks all in one place. I think I need to make sure to follow a similar pattern with `Transaction` so that it's less confusing, or at least that's what I'm trying first to see how it looks.